java/client/src/org/openqa/selenium/virtualauthenticator/Credential.java

```diff
@@ -0,0 +1,116 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.virtualauthenticator;
+
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A credential stored in a virtual authenticator.
+ * @see https://w3c.github.io/webauthn/#credential-parameters
+ */
+public class Credential {
+
+  private final byte[] id;
+  private final boolean isResidentCredential;
+  private final String rpId;
+  private final PKCS8EncodedKeySpec privateKey;
+  private final byte[] userHandle;
+  private final int signCount;
+
+  /**
+   * Creates a non resident (i.e. stateless) credential.
+   */
+  public static Credential createNonResidentCredential(byte[] id, String rpId,
+      PKCS8EncodedKeySpec privateKey, int signCount) {
+    return new Credential(id, /*isResidentCredential=*/false, Objects.requireNonNull(rpId),
+        privateKey, /*userHandle=*/null, signCount);
+  }
+
+  /**
+   * Creates a resident (i.e. stateful) credential.
+   */
+  public static Credential createResidentCredential(byte[] id, String rpId,
+      PKCS8EncodedKeySpec privateKey, byte[] userHandle, int signCount) {
+    return new Credential(id, /*isResidentCredential=*/true, Objects.requireNonNull(rpId),
+        privateKey, Objects.requireNonNull(userHandle), signCount);
+  }
+
+  /**
+   * Creates a credential from a map.
+   */
+  public static Credential fromMap(Map<String, Object> map) {
+    Base64.Decoder decoder = Base64.getUrlDecoder();
+    return new Credential(decoder.decode((String) map.get("credentialId")),
+        (boolean) map.get("isResidentCredential"),
+        (String) map.get("rpId"),
+        new PKCS8EncodedKeySpec(decoder.decode((String) map.get("privateKey"))),
+        map.get("userHandle") == null ? null : decoder.decode((String) map.get("userHandle")),
+        ((Long) map.get("signCount")).intValue());
+  }
+
+  private Credential(byte[] id, boolean isResidentCredential, String rpId,
+      PKCS8EncodedKeySpec privateKey, byte[] userHandle, int signCount) {
+    this.id = Objects.requireNonNull(id);
+    this.isResidentCredential = isResidentCredential;
+    this.rpId = rpId;
+    this.privateKey = Objects.requireNonNull(privateKey);
+    this.userHandle = userHandle;
+    this.signCount = signCount;
+  }
+
+  public byte[] getId() {
+    return id;
+  }
+
+  public boolean getIsResidentCredential() {
```

Comment:
Although it's not part of the javabean spec, it's legit to just use `is` as the prefix for boolean properties. I guess if the name of the property in the spec is "is resident credential" then `get` makes a lot of sense. Whichever you feel reads better and "feels" better is fine by me.

Reply:
I'll take any chance to make java less verbose. Changed to `isResidentCredential()`.